ext/strscan/strscan.c

```diff
@@ -714,14 +706,24 @@ strscan_do_scan(VALUE self, VALUE pattern, int succptr, int getstr, int headonly
         }
     }
     else {
+        StringValue(pattern);
         rb_enc_check(p->str, pattern);
         if (S_RESTLEN(p) < RSTRING_LEN(pattern)) {
             return Qnil;
         }
-        if (memcmp(CURPTR(p), RSTRING_PTR(pattern), RSTRING_LEN(pattern)) != 0) {
-            return Qnil;
+
+        if (headonly) {
+            if (memcmp(CURPTR(p), RSTRING_PTR(pattern), RSTRING_LEN(pattern)) != 0) {
+                return Qnil;
+            }
+            set_registers(p, RSTRING_LEN(pattern));
+        } else {
+            char *pos = strstr(CURPTR(p), RSTRING_PTR(pattern));
```

Comment:
Could you use `memmem()` instead of `strstr()`?
It seems that `StringValueCStr()` isn't called for `p->str`. (`p->str` may not be terminated with `\0` and `p-str` may include `\0`. For example: `StringScanner.new("a\0b").scan_until("b")`

Reply:
Oh.
I see.
Thanks!